RecommenderSystems/deepfm/deepfm_train_eval.py

```diff
@@ -0,0 +1,605 @@
+    state_dict = flow.load(f"{args.model_save_dir}/best_checkpoint", global_src_rank=0)
+    deepfm_module.load_state_dict(state_dict, strict=False)
+    print("================ Validation Evaluation ================")
+    eval(args, eval_graph, tag='val', cur_step=step, epoch=epoch, cached_eval_batches=cached_eval_batches)
+    print("================ Test Evaluation ================")
+    eval(args, eval_graph, tag='test', cur_step=step, epoch=epoch)
+
+
+def np_to_global(np, dtype=flow.float):
+    # TODO: t = flow.from_numpy(np)
+    t = flow.tensor(np, dtype=dtype)
+    return t.to_global(placement=flow.env.all_device_placement("cpu"), sbp=flow.sbp.split(0))
+
+
+def batch_to_global(np_label, np_features, is_train=True):
+    labels = np_to_global(np_label.reshape(-1, 1)) if is_train else np_label.reshape(-1, 1)
+    features = np_to_global(np_features, dtype=flow.int64)
+    return labels, features
+
+
+def prefetch_eval_batches(data_dir, batch_size, num_batches):
+    cached_eval_batches = []
+    with make_criteo_dataloader(data_dir, batch_size, shuffle=False) as loader:
+        for _ in range(num_batches):
+            label, features = batch_to_global(*next(loader), is_train=False)
+            cached_eval_batches.append((label, features))
+    return cached_eval_batches
+
+
+def eval(args, eval_graph, tag='val', cur_step=0, epoch=0, cached_eval_batches=None):
+    if tag == 'val':
+        batches_per_epoch = math.ceil(args.num_val_samples / args.batch_size)
+    else:
+        batches_per_epoch = math.ceil(args.num_test_samples / args.batch_size)
+    
+    eval_graph.module.eval()
+    labels, preds = [], []
+    eval_start_time = time.time()
+    
+    if cached_eval_batches == None:
+        with make_criteo_dataloader(f"{args.data_dir}/{tag}", args.batch_size, shuffle=False) as loader:
+            eval_start_time = time.time()
+            for i in range(batches_per_epoch):
+                label, features = batch_to_global(*next(loader), is_train=False)
+                pred = eval_graph(features)
+                labels.append(label)
+                preds.append(pred.to_local())
+    else:
+        for i in range(batches_per_epoch):
+            label, features = cached_eval_batches[i]
+            pred = eval_graph(features)
+            labels.append(label)
+            preds.append(pred.to_local())
+
+    labels = (
+        np_to_global(np.concatenate(labels, axis=0)).to_global(sbp=flow.sbp.broadcast()).to_local()
+    )
+    preds = (
+        flow.cat(preds, dim=0)
+        .to_global(placement=flow.env.all_device_placement("cpu"), sbp=flow.sbp.split(0))
+        .to_global(sbp=flow.sbp.broadcast())
+        .to_local()
+    )
+
+    flow.comm.barrier()
+    eval_time = time.time() - eval_start_time
+    
+    auc = 0  # will be updated by rank 0 only
+    logloss = 0 # will be updated by rank 0 only
+    rank = flow.env.get_rank()
+    if rank == 0:
```

Comment:
all devices should calculate auc and log_loss, but only rank 0 print loss and so on.